components/brave_shields/browser/tracking_protection_service.cc

```diff
@@ -49,6 +70,147 @@ TrackingProtectionService::~TrackingProtectionService() {
+  return std::tie(render_process_id, frame_routing_id) <
+         std::tie(other.render_process_id, other.frame_routing_id);
+}
+
+bool TrackingProtectionService::RenderFrameIdKey::operator==(
+    const RenderFrameIdKey& other) const {
+  return render_process_id == other.render_process_id &&
+         frame_routing_id == other.frame_routing_id;
+}
+
+void TrackingProtectionService::SetStartingSiteForRenderFrame(
+  GURL starting_site, int render_process_id, int render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey key(render_process_id, render_frame_id);
+  std::map<RenderFrameIdKey, GURL>::iterator iter =
+      render_frame_key_to_starting_site_url.find(key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    render_frame_key_to_starting_site_url.erase(key);
+  }
+  render_frame_key_to_starting_site_url.insert(std::pair<RenderFrameIdKey,
+    GURL>(key, starting_site));
+  return;
+}
+
+GURL TrackingProtectionService::GetStartingSiteForRenderFrame(
+  int render_process_id, int render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey key(render_process_id, render_frame_id);
+  std::map<RenderFrameIdKey, GURL>::iterator iter =
+      render_frame_key_to_starting_site_url.find(key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    return iter->second;
+  }
+  return GURL();
+}
+
+void TrackingProtectionService::ModifyRenderFrameKey(
+  int old_render_process_id, int old_render_frame_id,
+  int new_render_process_id, int new_render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey old_key(old_render_process_id, old_render_frame_id);
+  std::map<RenderFrameIdKey, GURL>::iterator iter =
+      render_frame_key_to_starting_site_url.find(old_key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    const RenderFrameIdKey new_key(new_render_process_id, new_render_frame_id);
+    render_frame_key_to_starting_site_url.insert(
+        std::pair<RenderFrameIdKey, GURL>(new_key, iter->second));
+    render_frame_key_to_starting_site_url.erase(old_key);
+  }
+}
+
+void TrackingProtectionService::DeleteRenderFrameKey(int render_process_id,
+  int render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey key(render_process_id, render_frame_id);
+  std::map<RenderFrameIdKey, GURL>::iterator iter =
+      render_frame_key_to_starting_site_url.find(key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    render_frame_key_to_starting_site_url.erase(key);
+  }
+}
+
+bool TrackingProtectionService::ShouldStoreState(HostContentSettingsMap* map,
+  int render_process_id, int render_frame_id, const GURL& top_origin_url,
+  const GURL& origin_url) {
+  const base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+  if (!command_line.HasSwitch(switches::kEnableSmartTrackingProtection)) {
```

Comment:
`base::Feature` ?

Reply:
@iefremov - I added this so we have a runtime-switch in `brave-browser` to enable/disable `smart-tracking-protection`. 

I'm not sure if there is a way to `enable` feature because we override the `enable-features` flag in `BraveMainDelegate`:  https://github.com/brave/brave-core/blob/adf9ea89fa654769104d649dbacc93fc4b1c760e/app/brave_main_delegate.cc#L129

May be we can add a switch that adds the `smart-tracking-protection` to enabled features? Let me know what you think.

Reply:
you can just declare your own feature (enabled or disabled as you wish) in the code (lots of examples in `*_features.h` files in the codebase). Then it will be possible to manipulate this feature by providing something like `--enable-features=SmartTrackingProtection`